### DISCUSSION TITLE
Add support for PICO TC08

### DISCUSSION THREAD

**Discussion by waanito** [artisan-roaster-scope/artisan#1035] [Ideas]
Hoping to add support for the Pico TC08 USB thermocouple data logger. 

I have a programming background in a variety of languages and have reverse engineered the omega MWTC wireless thermocouple  transmitter/receiver and written an application in visual studio c# to monitor coffee roasting in my RK drum using the MWTC.

With that background I'd like to have a go at developing a driver interface to artisan-scope for the tc08 and maybe later for the MWTC as well. Prefer to start with the TC08 as it has a better temperature resolution than the MWTC.

Should be a fun journey.
Feel free to discourage me

**Comment by poundy** [comment]
yeah don't do that. 

sorry, all I got :)

**Reply by MAKOMO** [reply]
Why not first write a small script that reads the data from the device and prints it to the console? Best use modern asyncio streams for this

https://docs.python.org/3/library/asyncio-stream.html

Once this is up, we can integrate this together.

**Reply by waanito** [reply]
on macos I've written a python script that gets data asynchronously from the TC08 and displays it to the console but it's via USB and the TC08 python wrapper API. You tell it how many reads you want and it fills a buffer with data that you dump to the screen when done.
I suspect maybe I may have misunderstood and that instead you'd want routines that just read a single temp datum and print that to the screen when requested? 
Not sure how to do that using asyncio since that all seems to be using network.

**Reply by MAKOMO** [reply]
Maybe you want to share your code to see what you have?

I see upcoming issues to integrate this cross-platform as some c-lib driver needs to be bundled in the builds.

**Reply by waanito** [reply]
I'm a complete neophyte at contributing code here on github. To share code I guess need create a gist and upload my code there. Public? and then report here when done?

**Reply by poundy** [reply]
if it's not too complex, drop files here (appending .txt to the end of the filename beforehand) in this discussion. If it's more complex, perhaps zipping it and adding it here (also with .txt) might do. But eventually you're going to want to have your own github setup where you can do pull requests from so it ends up being incorporated into Artisan's codebase.

**Reply by waanito** [reply]
thanks. Not too complex but since I'm already signed up at github may as well do it the right way. Be a day or two til I make it look better for public consumption and document stuff.

**Reply by waanito** [reply]
Thanks for your patience. OK after lots of debugging I have pushed 3 files to my github repo https://github.com/waanito/artisan-scope-tc08-macos
Took a while to get any pico sdk command at all running running on macos see the prerequisites file for what's needed.
The single mode example almost worked out of the box copied from picotech's github https://github.com/picotech just had to delete a non existent ref at end
The streaming mode example is what took me most time.  mostly trying to understand the the way the python wrapper functions work especially with arrays, and then expanding the python version to work like the c/c++ version they have. 
 
Now that I've got this running on macos I'd be happy to try for both linux and windows. 

Anyway let me know if there's any chance of adding the TC08 to artisan-scope on at least macos

**Reply by waanito** [reply]
I deleted those files and pushed a couple of updated files 
https://github.com/waanito/artisan-scope-tc08-macos

**Comment by waanito** [comment]
Thanks for that. Will do.

**Comment by waanito** [comment]
I've updated my github repo with the file to use the PICO TC08 with artisan-scope on macos
https://github.com/waanito/artisan-scope-tc08-macos

in a week or three I'll try to produce similar for windows and linux
The Omega MWTC  on the other hand is a poor candidate for use with artisan as it only does synchronized polling and its 2 second interval is slightly off from 2 secs but since that's what I'm using with my RK drum I may persist
